nomad/guests/base.py

```diff
@@ -0,0 +1,147 @@
+        # plugin system built in nomad.
+
+        return new_guest
+
+    def __init__(cls, name, bases, attrs):
+        # We implement the "mount point" paradigm here in order to make all `Guest` subclassses
+        # available in a single attribute.
+        if not hasattr(cls, 'guests'):
+            # The class has no guests list: this means that we are considering the "plugin mount"
+            # class. So we created the list that will hold all the defined `Guest` subclasses.
+            cls.guests = []
+        else:
+            # The `guests` attribute already exists so we are considering a `Guest` subclass, which
+            # needs to be registered.
+            cls.guests.append(cls)
+
+
+class Guest(with_metaclass(_GuestBase)):
+    """ Represents a single guest.
+
+    `Guest` subclasses will be used by `Container` instances to perform common operations on the
+    guest side. For example they can be used to ensure that some packages are installed when doing
+    barebone setups, to create some users, ... `Guest` subclasses should correspond to specific OSes
+    or distributions that can be runned as LXD containers.
+    """
+
+    # The `name` of a guest is a required attribute and should always be set on `Guest` subclasses.
+    name = None
+
+    # A list of packages that should be installed when doing barebone setup.
+    barebones_packages = []
+
+    def __init__(self, lxd_container):
+        self.lxd_container = lxd_container
+
+    @classmethod
+    def detect(cls, lxd_container):
+        """ Detects if the container is an "instance" of the considered OS.
+
+        The operations defined in this method assume that OS/distribution information will be
+        available through the analysis of files such as ""/etc/os-release", ""/etc/lsb_release" or
+        ""/etc/issue". This classmethod should be overriden if these methods cannot be applied to
+        the considered OS/distribution.
+        """
+        candidates = [
+            ('/etc/os-release', 'id="?{}"?'.format(cls.name.lower())),
+            ('/etc/lsb_release', 'id="?{}"?'.format(cls.name.lower())),
+            ('/etc/issue', cls.name.lower()),
+        ]
+        found = False
+        for candidate, pattern in candidates:
+            try:
+                candidate_filecontent = lxd_container.files.get(candidate)
+            except NotFound:
+                continue
+            found = len(re.findall(pattern, str(candidate_filecontent.lower()))) > 0
+            if found:
+                break
+        return found
+
+    def add_ssh_pubkey_to_root_authorized_keys(self, pubkey):
+        """ Add a given SSH public key to the root user's authorized keys. """
+        logger.info("Adding {} to machine's authorized keys".format(pubkey))
+        self.run(['mkdir', '-p', '/root/.ssh'])
+        self.lxd_container.files.put('/root/.ssh/authorized_keys', pubkey)
+
+    def create_user(self, username, home=None):
```

Comment:
What about making the default for `home` to be `/home/{ username }` (which is the most common case) rather than no home at all? We could support `home: false` too support cases where we want to create homeless users.